### PR TITLE
Change CPU sample sanitization in HPA.

### PR DESCRIPTION
**Pull request by krzysztof-jastrzebski**
**What this PR does / why we need it**:
Change CPU sample sanitization in HPA.
    Ignore samples if:
    - Pod is beeing initalized - 5 minutes from start defined by flag
        - pod is unready
        - pod is ready but full window of metric hasn't been colected since
        transition
    - Pod is initialized - 5 minutes from start defined by flag:
        - Pod has never been ready after initial readiness period.

**Release notes:**
```release-note
Improve CPU sample sanitization in HPA by taking metric's freshness into account.
```

### CODE REVIEW

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]
Couple of quick comments inline.  Need to do a full scan later today.

File: pkg/controller/podautoscaler/metrics/legacy_metrics_client.go [hunk 1]
```diff
@@ -35,10 +35,11 @@ import (
...
 const (
-	DefaultHeapsterNamespace = "kube-system"
-	DefaultHeapsterScheme    = "http"
-	DefaultHeapsterService   = "heapster"
-	DefaultHeapsterPort      = "" // use the first exposed port on the service
+	DefaultHeapsterNamespace    = "kube-system"
+	DefaultHeapsterScheme       = "http"
+	DefaultHeapsterService      = "heapster"
+	DefaultHeapsterPort         = "" // use the first exposed port on the service
+	heapsterdefaultMetricWindow = time.Minute
```
[the remark this answers is not in the document]
Comment (krzysztof-jastrzebski) [reply pkg/controller/podautoscaler/metrics/legacy_metrics_client.go]: Fixed

File: cmd/kube-controller-manager/app/options/options.go [hunk 2]
```diff
@@ -134,7 +135,7 @@ func NewKubeControllerManagerOptions() (*KubeControllerManagerOptions, error) {
 			HorizontalPodAutoscalerSyncPeriod:               componentConfig.HPAController.HorizontalPodAutoscalerSyncPeriod,
 			HorizontalPodAutoscalerUpscaleForbiddenWindow:   componentConfig.HPAController.HorizontalPodAutoscalerUpscaleForbiddenWindow,
 			HorizontalPodAutoscalerDownscaleForbiddenWindow: componentConfig.HPAController.HorizontalPodAutoscalerDownscaleForbiddenWindow,
-			HorizontalPodAutoscalerCPUTaintPeriod:           componentConfig.HPAController.HorizontalPodAutoscalerCPUTaintPeriod,
+			HorizontalPodAutoscalerCPUInitializationPeriod:  componentConfig.HPAController.HorizontalPodAutoscalerCPUInitializationPeriod,
```
[the remark this answers is not in the document]
Comment (krzysztof-jastrzebski) [reply cmd/kube-controller-manager/app/options/options.go]: +mwielgus
Currently this is only for CPU and I belive it will be only used for CPU.

[the remark this answers is not in the document]
Comment (DirectXMan12) [reply cmd/kube-controller-manager/app/options/options.go]: it got changed a PR or two ago.  It used to be universal across all metrics, so we probably need to decide if we really want to change it.

[the remark this answers is not in the document]
Comment (DirectXMan12) [reply cmd/kube-controller-manager/app/options/options.go]: @mwielgus

File: pkg/controller/podautoscaler/metrics/legacy_metrics_client_test.go [hunk 3]
```diff
@@ -364,3 +404,7 @@ func testCollapseTimeSamples(t *testing.T) {
 	assert.InEpsilon(t, float64(75), val, 0.1, "collapsed sample value should be as expected")
 	assert.True(t, timestamp.Equal(now), "timestamp should be the current time (the newest)")
 }
+
+func calcTimestamp(t int) time.Time {
```
Comment (DirectXMan12) [inline pkg/controller/podautoscaler/metrics/legacy_metrics_client_test.go]: this should have a way better name.  These tests are already hard to follow (maybe call it `offsetTimestampBy` or something).

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]
minor nit inline, otherwise fine modulo the conversation about where this applies

File: cmd/kube-controller-manager/app/options/options.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (krzysztof-jastrzebski) [reply cmd/kube-controller-manager/app/options/options.go]: We didn't have such flag before and we want to change behavior only for CPU as it is kind of special.

**Comment by krzysztof-jastrzebski** [comment]
Is there anything else that I should fix apart calcTimestamp name?

**Comment by DirectXMan12** [comment]
I think I'm fine with this otherwise.

File: pkg/controller/podautoscaler/metrics/legacy_metrics_client_test.go (hunk 3 above)
Comment (krzysztof-jastrzebski) [reply pkg/controller/podautoscaler/metrics/legacy_metrics_client_test.go]: Fixed.

**Comment by krzysztof-jastrzebski** [comment]
Please give me LGTM so I can collect approvals tomorrow.
